File: src/Python/221.MaximalSquare.py

class Solution:
    def maximalSquare(self, matrix):
        rows = len(matrix)
        if rows == 0:
            return 0
        cols = len(matrix[0])
        dp = [0] * (cols + 1)
        maxsqlen = 0
        prev = 0
        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                temp = dp[j]
                if matrix[i - 1][j - 1] == '1':
                    dp[j] = min(min(dp[j - 1], prev), dp[j]) + 1
                    maxsqlen = max(maxsqlen, dp[j])
                else:
                    dp[j] = 0
                prev = temp
        return maxsqlen * maxsqlen


if __name__ == "__main__":
    solution = Solution()
    print(solution.maximalSquare([['1','0','1','0','0'],['1','0','1','1','1'],['1','1','1','1','1'],['1','0','0','1','0']]))

File: src/CSharp/221.MaximalSquare.cs

using System;
using System.Collections.Generic;

namespace PeefyLeetCode.MaximalSquare
{
public class Solution {
    public int MaximalSquare(char[][] matrix) {
        int rows = matrix.Length;
        if (rows == 0)
            return 0;
        int cols = matrix[0].Length;
        int[] dp = new int[cols + 1];
        int prev = 0;
        int maxn = 0;
        for (int i = 1; i <= rows; i++) {
            for (int j = 1; j <= cols; j++) {
                int temp = dp[j];
                if (matrix[i - 1][j - 1] == '1'){
                    dp[j] = Math.Min(Math.Min(dp[j - 1], prev), dp[j]) + 1; 
                    maxn = Math.Max(maxn, dp[j]);
                }
                else
                    dp[j] = 0;
                prev = temp;
            }
        }
        return maxn * maxn;
    }
}
}